PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h

```diff
@@ -271,3 +272,134 @@ class FirstObjectSimpleFlatTableProducer : public SimpleFlatTableProducerBase<T,
     return out;
   }
 };
+
+template <typename T, typename TProd>
+class SimpleFlatTableProducerBaseLumi : public edm::one::EDProducer<edm::EndLuminosityBlockProducer> {
```

Comment:
Why does this module (template) need to be `edm::one`? Could it be `edm::stream`?

If it needs to stay `edm::one` for some reason, please use also [`edm::LuminosityBlockCache`](https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkGlobalModuleInterface#edm_LuminosityBlockCacheT) extension to tell the framework the the module can process events from multiple LuminosityBlocks concurrently.

Reply:
Dear @makortel,

At the time of development I tried to use the `edm::stream` (as in _SimpleFlatTableProducerBase_) guided by the [_FWMultithreadedFrameworkStreamModuleInterface_ Twiki](https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkStreamModuleInterface#edm_BeginLuminosityBlockProducer) however, this resulted in a series of compilation errors that I was not able to resolve. Using `edm::one` solved the issues.

If you have any suggestions on how to make `edm::stream` work, I would be grateful.

Best,
Mateusz

Reply:
What kind of compilation errors did you get?

Reply:
The first one is: `/afs/cern.ch/work/m/mzarucki/nanoAOD/CMSSW_12_0_0_pre3/src/PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h:315:8: error: 'void SimpleFlatTableProducerBaseLumi<T, TProd>::endLuminosityBlockProduce(edm::LuminosityBlock&, const edm::EventSetup&) [with T = GenFilterInfo; TProd = GenFilterInfo]' marked 'final', but is not virtual`

which can be solved by just removing the `final` specifier for the `endLuminosityBlockProduce` method. 

A following compilation error is:

`/cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/src/FWCore/Framework/interface/stream/callAbilities.h:455:43: error: 'globalEndLuminosityBlockProduce' is not a member of 'LumiSingletonSimpleFlatTableProducer<GenFilterInfo>'
  455 |         T::globalEndLuminosityBlockProduce(Lumi, iES, iRC);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~`

Reply:
My bad, after further thought stream module type actually wouldn't help. I was thinking we could use the "streams" to work around the thread-unsafety of `StringObjectFunction` and `StringCutObjectSelector`, but the LuminosityBlock transitions of stream modules are global (i.e. can occur concurrently).

So the best way to support "processing Events from multiple LuminosityBlocks concurrently" would be to use the ` edm::LuminosityBlockCache` extension. Since you don't really need the cache, you could declare it as e.g. `edm::LuminosityBlockCache<int>` and in `globalBeginLuminosityBlock()` return `nullptr`.

Reply:
Dear @makortel,

Therefore, is `edm:one` fine to use for this module?

As suggested, I have added the `edm:LuminosityBlockCache` extension as follows: 
```
class SimpleFlatTableProducerBaseLumi : public edm::one::EDProducer<edm::EndLuminosityBlockProducer, edm::LuminosityBlockCache<int>> {
```

I have also added the `globalBeginLuminosityBlock` method which returns `nullptr`.

Unfortunately, I get the following compilation errors:
```
 from /afs/cern.ch/work/m/mzarucki/nanoAOD/CMSSW_12_0_0_pre3/src/PhysicsTools/NanoAOD/plugins/SimpleFlatTableProducerPlugins.cc:1:
/cvmfs/cms.cern.ch/slc7_amd64_gcc900/external/gcc/9.3.0/include/c++/9.3.0/bits/unique_ptr.h: In instantiation of 'typename std::_MakeUniq<_Tp>::__single_object std::make_unique(_Args&& ...) [with _Tp = LumiSingletonSimpleFlatTableProducer<GenFilterInfo>; _Args = {const edm::ParameterSet&}; typename std::_MakeUniq<_Tp>::__single_object = std::unique_ptr<LumiSingletonSimpleFlatTableProducer<GenFilterInfo>, std::default_delete<LumiSingletonSimpleFlatTableProducer<GenFilterInfo> > >]':
/cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/src/FWCore/Framework/src/MakeModuleHelper.h:39:40:   required from 'static std::unique_ptr<_Tp> edm::MakeModuleHelper<Base>::makeModule(const edm::ParameterSet&) [with T = LumiSingletonSimpleFlatTableProducer<GenFilterInfo>; Base = edm::one::EDProducerBase]'
/cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/src/FWCore/Framework/src/WorkerMaker.h:83:107:   required from 'std::shared_ptr<edm::maker::ModuleHolder> edm::WorkerMaker<T>::makeModule(const edm::ParameterSet&) const [with T = LumiSingletonSimpleFlatTableProducer<GenFilterInfo>]'
/cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/src/FWCore/Framework/src/WorkerMaker.h:77:40:   required from here
/cvmfs/cms.cern.ch/slc7_amd64_gcc900/external/gcc/9.3.0/include/c++/9.3.0/bits/unique_ptr.h:857:30: error: invalid new-expression of abstract class type 'LumiSingletonSimpleFlatTableProducer<GenFilterInfo>'
  857 |     { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Is there something obvious that I might be missing here?

Best regards,
Mateusz

Reply:
The use of `LuminosityBlockCache` requires overriding also `void SimpleFlatTableProducerBaseLumi::globalEndLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) override {}`.
(sorry for not mentioning it earlier)
https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkOneModuleInterface#edm_LuminosityBlockCacheT

Reply:
Thanks for the clarification!

I have made the discussed updates to support processing events from multiple LuminosityBlocks concurrently in https://github.com/cms-sw/cmssw/pull/34169/commits/a4a7f44db5a06f68f25b308f0d0e1b0e8534bb06 

I would be grateful if you could have a look whether this is what you were expecting.

Cheers,
Mateusz

Reply:
> I would be grateful if you could have a look whether this is what you were expecting.

Yes, it looks good now. Thanks!